Documentation for rc0 tasks (#2819)
kubernetes/sig-release: release-engineering/handbooks/post-release-branch-creation.md

- Update version references
- Rotate job configurations (n -> n+1)
- Add new version as beta
- Update stable versions sequentially
- Update configuration for all test suites

> [!NOTE]
Just shift "args" but not touch interval, sigowners or any other field.
Remember that args are bound to a set of tests for a specific release, while intervals and everything else is bound to a release "status": stable1/2/3/4 or beta. An example of how to correctly rotate the jobs can be found [here](https://github.com/kubernetes/test-infra/pull/34668/commits/819c9d253ab873aff6626a5eaf7635560f7b769e).

The docs related to these steps are available in the [test-infra releng README](https://github.com/kubernetes/test-infra/blob/master/releng/README.md).
Below is a more verbose breakdown of the steps.

#### Create the release dashboards

After running the make command you can now update the [release dashboards](https://github.com/kubernetes/test-infra/blob/master/config/testgrids/kubernetes/sig-release/config.yaml), example [commit](https://github.com/kubernetes/test-infra/commit/31fb8f2b5c4458af675e37765dfebd128da19971), remembering to:

> - Remove the deprecated release sig-release-1.30-{blocking,informing} dashboards

Author: I think a second iteration is needed to have a separate doc for removing EOL jobs, agreed

#### Update k8s-ci-builder image

The k8s-ci-builder needs to be updated in a similar fashion:

1. Create a PR to update the k8s-ci-builder variants.yaml file:
   - Edit [images/releng/k8s-ci-builder/variants.yaml](https://github.com/kubernetes/release/blob/master/images/releng/k8s-ci-builder/variants.yaml)
   - Add a new variant with the appropriate Go version that matches the release

   ```yaml
   '1.34':
      CONFIG: '1.34'
      GO_VERSION: '1.24.5'
      GO_VERSION_TOOLING: '1.24.5'
      OS_CODENAME: 'bullseye'
   ```

2. This can be included in the same PR as the k8s-cloud-builder update
   - If submitted separately, follow the same process as the k8s-cloud-builder update.

After all the above PRs are merged and prow jobs are successful, you can verify the images were built successfully:

   ```bash
   # Check the kube-cross image
   gcloud container images list-tags gcr.io/k8s-staging-build-image/kube-cross

   # Check the k8s-cloud-builder image
   gcloud container images list-tags gcr.io/k8s-staging-releng/k8s-cloud-builder

   # Check the k8s-ci-builder image
   gcloud container images list-tags gcr.io/k8s-staging-releng/k8s-ci-builder
   ```

> ### Update k8s references for kube-cross image

Reviewer: ```suggestion
### Update `kubernetes/kubernetes` references for the `kube-cross` image
```

Once the new builder images are available:

Create a PR in kubernetes/kubernetes to reference the new kube-cross image in `build/build-image/cross/VERSION`:

>    - Example PR: https://github.com/kubernetes/kubernetes/pull/132897/files

Reviewer: ```suggestion
   - Example PR: https://github.com/kubernetes/kubernetes/pull/132897
```

The Kubernetes Publishing Bot is responsible for:
- ensuring that the master and release branches in the staging repositories are in-sync with the appropriate branches in `kubernetes/kubernetes`
- creating tags in the staging repositories for each Kubernetes release

Update the publishing bot rules to include the new version. Use the update-rules CLI tool to generate the necessary changes as described [here](https://github.com/kubernetes/publishing-bot/blob/master/cmd/update-rules/README.md).

You can install the tool on Linux with a simple `make build` command, similarly on MacOS by specifying your OS `GOOS=darwin make build`.
By default the binary will be located in `_output/update-rules`.

You can also use the Docker image to run the tool, which is useful if you don't want to build it locally (but you can also do that with make `make build-image`). The live Docker image is available at `gcr.io/k8s-staging-publishing-bot/publishing-bot:latest` and you can invoke the tool as follows:

```bash
docker run -t gcr.io/k8s-staging-publishing-bot/publishing-bot:latest /update-rules -branch release-1.34 -go 1.24.5 -rules ~/kubernetes/staging/publishing/rules.yaml -o /tmp/rules.yaml
```

> Recommended to set the release branch and the Go version as environment variables before running the command:

Reviewer: ```suggestion
It's recommended to set the release branch and the Go version as environment variables before running the command:
```

They must be executed after the `nomock release` is completed, and the release branch is created, as stated in the [branch creation chapter](k8s-release-cut.md#next-release-branch-creation) of the release cut handbook.

PR can be created beforehand (and this is recommended in order to get reviews in a timely manner) but you got to remember to put a `/hold` on all the PRs, they have to be lifted only once the `nomock` release phase is done and the branch is created.

> [!WARNING]
It is essential to follow these steps to maintain the integrity of the release process and ensure that all necessary components are updated accordingly, the examples provided are for illustrative purposes and need to be adapted to the specific release version, so please replace `1.30`, `1.33` and `1.34` with the actual version(s) you are deprecating, currently working on or preparing grounds for.

## Checklist

Open a new issue using [this template](https://github.com/kubernetes/sig-release/issues/new?template=post-release-branch-creation.md).

### Promote the subsequent minor release alpha.0 image

Recall that an alpha.0 of the next minor release is created during the rc.0 cut?

> To assist downstream consumers of Kubernetes, this new alpha is automatically cut to bring our next release tag to the tip of `master`, what we need to do manually is [promoting](k8s-release-cut.md#7-image-promotion) the alpha.0 image, after the rc.0 image ProwJob is green.

Reviewer: This document is not really appropriate for this task. You'd do this much before you get to this document, i.e. before even running `nomock release` for rc.0.

PR can be created beforehand (and this is recommended in order to get reviews in a timely manner) but you got to remember to put a `/hold` on all the PRs, they have to be lifted only once the `nomock` release phase is done and the branch is created.

> [!WARNING]
It is essential to follow these steps to maintain the integrity of the release process and ensure that all necessary components are updated accordingly, the examples provided are for illustrative purposes and need to be adapted to the specific release version, so please replace `1.30`, `1.33` and `1.34` with the actual version(s) you are deprecating, currently working on or preparing grounds for.

## Checklist

Open a new issue using [this template](https://github.com/kubernetes/sig-release/issues/new?template=post-release-branch-creation.md).

### Cut next alpha

Recall that an alpha.0 of the next minor release was created during the rc.0 cut?

To assist downstream consumers of Kubernetes, a new alpha must be cut to bring our next release tag to the tip of `master`.

Begin the release process following the [release cut handbook](k8s-release-cut.md) following the instructions tailored for alpha releases.

> This needs to be done before the `nomock release` command is executed for the rc.0.

Reviewer: This sounds wrong to me. This should eventually be done after the Code Thaw, but definitely not within the post-branch creation tasks (assuming alpha.0 is automatic and alpha.1 after the .0 release).

> ### Remove EOL version jobs from test-infra (optional)

Reviewer: As mentioned above, I'd kick this out from the post-branch creation tasks.

Author: As above, I think keeping it is right due to how connected this is to the success of the post branch creation tasks.

If you don't remove the old jobs the test cluster will be overloaded and flakiness would be up, at the very least.

# Changes would involve removing the EOL version from: images/kubekins-e2e-v2/variants.yaml

# Diff example:

- '1.30':
-    CONFIG: '1.30'
-    GO_VERSION: 1.23.6
-    K8S_RELEASE: latest-1.30
-    BAZEL_VERSION: 3.4.1
```

> [!NOTE]
This step may alternatively be performed as part of a patch release process. It's mandatory to consult with the release engineering team regarding the timing of this step.

### Update milestone applier rules and check milestone requirements

Create a PR to update the milestone applier rules to include the new version and remove the oldest version.

> [!WARNING]
Only remove the oldest version if it is already EOL and the release branch jobs have been already removed.

```yaml
# Example PR: https://github.com/kubernetes/test-infra/pull/34650
# Edit this file: config/prow/plugins.yaml

# Add the new version and remove the oldest version in the kubernetes/kubernetes section - if the oldest version is EOL:
milestone_applier:
  kubernetes/kubernetes:
    master: v1.33
    release-1.33: v1.33
    release-1.32: v1.32
    ...
    # remove oldest version (e.g., release-1.29: v1.29)
```

> [!NOTE]

> Look out for the code freeze config and ensure excluded and included branches include the upcoming release branch `release-1.xy`.

Reviewer: Nit: the branch is already created at this point.
```suggestion
Look out for the code freeze config and ensure excluded and included branches include the newly created release branch `release-1.xy`.
```

### Update release branch jobs in kubernetes/test-infra for the new release and create the dashboards

> [!CAUTION]
Follow the guidelines below very carefully during the update process.

- Do not remove old jobs while adding new jobs, just do not. Let Release Engineering handle it before or after the post branch creation tasks.
- Do not segregate PRs, just separate auto-generated files from manually updated ones in two (or more) clearly documented commits.
- Be super careful about `releng/test_config.yaml` epecially when commenting out `stable4`
- Do not hesitate to remove broken jobs, but let the interested SIG(s) know about it so they can re-add it.

Updating the release branch jobs in `kubernetes/test-infra` for the new release version involves some steps, it is unpredictable work and requires manual intervention as not every job is generated.
Some of them are added manually and live outside of the generated job tree.

First of all you need to modify the `releng/test_config.yaml` file ([here](https://github.com/kubernetes/test-infra/blob/master/releng/test_config.yaml)) to:
- Update version references

> - Rotate stable job configurations (n -> n+1)

Author: Merged

This is related to how the tags are managed and you should consider using `latest` images instead or you will find yourself opening PRs like [this](https://github.com/kubernetes/test-infra/pull/35288) to fix the image tags for some jobs.

Remember to update all configs before running the [generation script](#run-test-generation-script) for the upcoming release branch jobs and to verify that the jobs are generated correctly.

#### Create the release dashboards

After configuring the jobs you can now update the [release dashboards](https://github.com/kubernetes/test-infra/blob/master/config/testgrids/kubernetes/sig-release/config.yaml), example [commit](https://github.com/kubernetes/test-infra/commit/31fb8f2b5c4458af675e37765dfebd128da19971), remembering to:

- Remove the deprecated release sig-release-1.xx-{blocking,informing} dashboards
- Add the new dashboards for the current release e.g., sig-release-1.xx-{blocking,informing}

> - Pairing with the Release Signal lead, check that the new dashboards are working by visiting the TestGrid pages for the new dashboards and verifying that all the necessary jobs show up correctly.

Author: Done and also reorganized this bit to be closer to when we actually mention the PR being opened and merging!
